### PR TITLE
pkg/derivation: Fix assumptions around name for structured attrs

### PR DESCRIPTION
**Pull request by adisbladis**
This PR in it's current form is more to demonstrate the issue than to propose an actual fix.

Depends on https://github.com/nix-community/go-nix/pull/70.

### CODE REVIEW

File: pkg/derivation/derivation.go [hunk 1]
```diff
@@ -122,6 +123,26 @@ func (d *Derivation) Validate() error {
 		if k == "name" {
 			hasNameEnv = true
 		}
+
+		// Structured don't have the env name right in the regular spot but in the nested JSON object.
+		// I don't particularly like to parse this object as a validation step so maybe we can just the name check
+		// if a derivation uses structured attrs.
```
Comment (flokli) [inline pkg/derivation/derivation.go:129]:
```suggestion
		// I don't particularly like to parse this object as a validation step so maybe 
		// we can only do the name check if a derivation uses structured attrs.
```

File: pkg/derivation/derivation.go [hunk 2]
```diff
@@ -122,6 +123,26 @@ func (d *Derivation) Validate() error {
...
+			type structuredAttrs struct {
+				Name string
+			}
+
+			attrs := &structuredAttrs{}
+
+			err := json.Unmarshal([]byte(d.Env[k]), &attrs)
+			if err != nil {
+				return err
+			}
```
Comment (flokli) [inline pkg/derivation/derivation.go:140]: Wrap the error with a useful error message what went wrong here.

File: pkg/derivation/drv_path.go [hunk 3]
```diff
@@ -54,10 +55,26 @@ func (d *Derivation) DrvPath() (string, error) {
...
+		type structuredAttrs struct {
+			Name string
+		}
+
+		attrs := &structuredAttrs{}
+
+		err := json.Unmarshal([]byte(structuredJson), &attrs)
+		if err != nil {
+			return "", err
+		}
```
Comment (flokli) [inline pkg/derivation/drv_path.go:69]: Wrap the error with a useful error message what went wrong.

File: pkg/derivation/hashes.go [hunk 4]
```diff
@@ -36,9 +37,27 @@ func (d *Derivation) getMaskedATermHash(inputDrvReplacements map[string]string)
...
+		type structuredAttrs struct {
+			Name string
+		}
+
+		attrs := &structuredAttrs{}
+
+		err := json.Unmarshal([]byte(structuredJson), &attrs)
+		if err != nil {
+			return nil, err
+		}
```
Comment (flokli) [inline pkg/derivation/hashes.go:52]: same

**Comment by flokli** [comment]
It's terrible we need to do this, but I think we should - and definitely need to do while calculating the derivation path.

Maybe we can extract the "get name of a derivation" part into a private function for now?

**Comment by flokli** [comment]
@adisbladis are you fine with merging this for now?

We can rework this later if we come up with a better way to convey this, but this should fix some of the hash calculations for now.

**Comment by adisbladis** [comment]
Yes this is fine for now.

**Pull request merged**